### PR TITLE
Proposal for the implementation of Popovers in UI-Service

### PR DESCRIPTION
**Pull request by wanze**
The implementation uses the same pattern (Triggerer/Signals/Triggerable) as introduced with the modals. Here's a short example how a component acting as triggerer can display a popover on hover:

```php
$popover = $factory->popover('Title', 'Hello World');
$button = $factory->button()->standard('Show Popover', '#')
    ->withOnHover($popover->getShowSignal());
```
Cheers

### CODE REVIEW

File: src/UI/Component/Popover/Popover.php [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+	 * @return string
+	 */
+	public function getText();
+
+	/**
+	 * Get the position of the popover: top, bottom, left, right or auto
+	 *
+	 * @return string
+	 */
+	public function getPosition();
```
Comment (klees) [inline src/UI/Component/Popover/Popover.php:32]: Is this really something the dev should decide? Or could we derive this from other information, maybe even client side?

File: src/UI/Implementation/Component/Popover/Renderer.php [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+		return '';
+	}
+
+
+	public function registerResources(ResourceRegistry $registry) {
+		parent::registerResources($registry);
+		$registry->register('./src/UI/templates/js/Popover/popover.js');
+	}
+
+	protected function esacpe($str) {
```
Comment (klees) [inline src/UI/Implementation/Component/Popover/Renderer.php]: `esacpe` -> `escape`

**Review by klees — COMMENTED** [review]
LGTM, thx!

File: src/UI/Implementation/Component/Popover/Renderer.php (hunk 2 above)
Comment (wanze) [reply src/UI/Implementation/Component/Popover/Renderer.php]: omg thanks

File: src/UI/Component/Popover/Popover.php (hunk 1 above)
Comment (wanze) [reply src/UI/Component/Popover/Popover.php:32]: The default position is set to `auto`, meaning that bootstrap is deciding where to display the popover based on the available space.

Comment (klees) [reply src/UI/Component/Popover/Popover.php:32]: Unless someone has a use case where its definitely necessary that the Dev decides, I would just drop this position property and always set it to auto. I currently can't see that use case.

Comment (Amstutz) [reply src/UI/Component/Popover/Popover.php:32]: The Dev. should decide as little as possible. We could talk about giving different popover instances and let the dev choose one (e.g. Top Navigation Popover, which always displays on the bottom of the item). I am not sure if this is needed. Needs to be discussed. The placement option of bootstrap is probably to limited for us. We should think about extending that a little. Similar to: [http://sandywalker.github.io/webui-popover/demo/](http://sandywalker.github.io/webui-popover/demo/)

File: src/UI/Component/Popover/Popover.php [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+	 * @return string
+	 */
+	public function getTitle();
+
+	/**
+	 * Get the text (content) of the popover
+	 *
+	 * @return string
+	 */
+	public function getText();
```
Comment (Amstutz) [inline src/UI/Component/Popover/Popover.php:25]: We might have a type of popover only accepting text. However there are certainly others as well. Not sure if this is placed at the correctly here.

File: src/UI/Factory.php [hunk 4]
```diff
@@ -336,4 +336,28 @@ public function panel();
 	 **/
 	public function modal();
 
+	/**
+	 * ---
+	 * description:
+	 *   purpose: >
+	 *     Popovers are used to display secondary information to a component.
```
Comment (Amstutz) [inline src/UI/Factory.php:343]: Not only. They might also contain interaction. Please specifiy further. See: [https://docs.google.com/document/d/1JoInU0-tkRHDOMjMTNziV-hJb6ypDCI1zT9OGtAThZk/edit#heading=h.um0atzynczok](https://docs.google.com/document/d/1JoInU0-tkRHDOMjMTNziV-hJb6ypDCI1zT9OGtAThZk/edit#heading=h.um0atzynczok)
However, note that the current description given in the pipeline document was rejected by the JF and therefore needs some revision.

File: src/UI/Factory.php [hunk 5]
```diff
@@ -336,4 +336,28 @@ public function panel();
...
+	 *   usage:
+	 *     1: Popovers MUST contain a title and a text.
+	 *     2: The text MUST be plain text (no HTML) and no longer than a paragraph.
+	 * ---
+	 * @param string $title
+	 * @param string $text
+	 * @param string $position
+	 * @return \ILIAS\UI\Component\Popover\Popover
+	 */
+	public function popover($title, $text, $position = 'auto');
```
Comment (Amstutz) [inline src/UI/Factory.php:361]: There will be certainly more than one popover type. Position should be moved out of the constructor. The constructor should be as small as possible.

File: src/UI/Factory.php [hunk 6]
```diff
@@ -336,4 +336,28 @@ public function panel();
 	 **/
 	public function modal();
 
+	/**
+	 * ---
+	 * description:
+	 *   purpose: >
+	 *     Popovers are used to display secondary information to a component.
+	 *   composition: >
+	 *     Popovers consist of a short title and text.
```
Comment (Amstutz) [inline src/UI/Factory.php:345]: Incomplete

File: src/UI/Factory.php [hunk 7]
```diff
@@ -336,4 +336,28 @@ public function panel();
...
+	 *   effect: >
+	 *     Popovers are typically shown by clicking or hovering over a component. The position
+	 *     of the Popover is calculated automatically be default. However, it is possible to
+	 *     specify if the popover appears at the top, right, bottom or left of the triggerer component.
+	 * context: >
+	 * rules:
+	 *   usage:
+	 *     1: Popovers MUST contain a title and a text.
+	 *     2: The text MUST be plain text (no HTML) and no longer than a paragraph.
```
Comment (Amstutz) [inline src/UI/Factory.php:354]: Very incomplete. Accessability is a huge topic in the case of popovers. Please checkout the rules proposed at: [https://docs.google.com/document/d/1JoInU0-tkRHDOMjMTNziV-hJb6ypDCI1zT9OGtAThZk/edit#heading=h.um0atzynczok](https://docs.google.com/document/d/1JoInU0-tkRHDOMjMTNziV-hJb6ypDCI1zT9OGtAThZk/edit#heading=h.um0atzynczok)

File: src/UI/Implementation/Component/Popover/Popover.php [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+
+class Popover implements Component\Popover\Popover {
+
+	use ComponentHelper;
+
+	const POS_LEFT = 'left';
+	const POS_BOTTOM = 'bottom';
+	const POS_TOP = 'top';
+	const POS_RIGHT = 'right';
+	const POS_AUTO = 'auto';
```
Comment (Amstutz) [inline src/UI/Implementation/Component/Popover/Popover.php:17]: We probably need to extend that with something like bottom-corner (think about our top navigation).

File: src/UI/templates/js/Popover/popover.js [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+
+        var toggle = function ($triggerer, options) {
+            options = JSON.parse(options);
+            $triggerer.popover(options).popover('toggle');
+        };
+
+        return {
+            toggle: toggle
+        };
+
```
Comment (Amstutz) [inline src/UI/templates/js/Popover/popover.js:16]: Think about adding further placement options here.

**Review by Amstutz — CHANGES_REQUESTED** [review]
Popovers seem to be not so much a technical challenge (except for the placement) as much more a semantical. What types of popovers are needed? What rules restrictions will apply to them? Here much more work is still needed. We will discuss much of it in the scheduled meeting, today 10:15. As always, everybody is invited to join.

**Comment by wanze** [comment]
Please see the follow-up pull request: https://github.com/ILIAS-eLearning/ILIAS/pull/513

**Pull request closed without merging**